src/hmem_ze.c

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright (c) 2020 Intel Corporation. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#if HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include "ofi_hmem.h"
+#include "ofi.h"
+
+#ifdef HAVE_LIBZE
+
+#include <level_zero/ze_api.h>
+
+#define ZE_MAX_DEVICES 4
+
+static ze_driver_handle_t driver;
+static ze_device_handle_t devices[ZE_MAX_DEVICES];
+static ze_command_queue_handle_t cmd_queue[ZE_MAX_DEVICES];
+static int num_devices = 0;
+
+static inline int _ze_cmd_queue_create(ze_device_handle_t device,
+				       ze_command_queue_handle_t *cmd_queue)
+{
+	ze_command_queue_desc_t desc = {
+		.version	= ZE_COMMAND_QUEUE_DESC_VERSION_CURRENT,
+		.flags		= ZE_COMMAND_QUEUE_FLAG_NONE,
+		.mode		= ZE_COMMAND_QUEUE_MODE_SYNCHRONOUS,
+		.priority	= ZE_COMMAND_QUEUE_PRIORITY_NORMAL,
+		.ordinal	= 0,
+	};
+
+	return zeCommandQueueCreate(device, &desc, cmd_queue);
+}
+
+static inline int _ze_cmd_list_create(ze_device_handle_t device,
+				      ze_command_list_handle_t *cmd_list)
+{
+	ze_command_list_desc_t desc = {
+		.version	= ZE_COMMAND_LIST_DESC_VERSION_CURRENT,
+		.flags		= ZE_COMMAND_LIST_FLAG_NONE,
+	};
+	return zeCommandListCreate(device, &desc, cmd_list);
+}
```

Comment:
Both of the above static inline calls wrap a single function and are only called from one place.  They don't really seem useful, and in the queue_create case, it results in initializing desc repeatedly in a loop.